lib/execution_segment.go

```diff
@@ -394,13 +395,151 @@ func (ess ExecutionSegmentSequence) String() string {
+	// No filling up is required as the algorithm will accommodate for it
+	// through just going through the iterations that need to be in the values will fill up
+	// this has the consequence that if this is ran without sequence,
+	// but with segments: 0:1/3 and 1/3:2/3 it will get the same results instead
+	// of 1/3:2/3 to get start=1 and offset={3} it will get as 0:1/3 will start=0 and offsets={3}
+	// if the above behaviour is desired this will definitely need to be outside of this function.
+	if ess == nil || len(*ess) == 0 {
+		copyESS = []*ExecutionSegment{segment}
+	} else {
+		copyESS = append([]*ExecutionSegment{}, *ess...) // copy the original sequence
+	}
+	var wrapper = newWrapper(copyESS)
+
+	var segmentIndex = wrapper.indexOf(segment)
+	if segmentIndex == -1 {
+		return -1, nil, -1, fmt.Errorf("missing segment %s inside segment sequence %s", segment, ess)
+	}
+	start, offsets := wrapper.strippedOffsetsFor(segmentIndex)
+	return start, offsets, wrapper.lcd, nil
+}
+
+// This is only needed in order to sort all three at the same time
+type sortInterfaceWrapper struct { // TODO: rename ?
+	ess        ExecutionSegmentSequence
+	numerators []int64
+	lcd        int64
+}
+
+func newWrapper(ess ExecutionSegmentSequence) sortInterfaceWrapper {
+	var result = sortInterfaceWrapper{
+		ess:        ess,
+		numerators: make([]int64, len(ess)),
+		lcd:        ess.lcd(),
+	}
+
+	for i := range ess {
+		result.numerators[i] = ess[i].length.Num().Int64() * (result.lcd / ess[i].length.Denom().Int64())
+	}
+
+	sort.Stable(result)
+	return result
+}
+
+func (e sortInterfaceWrapper) indexOf(segment *ExecutionSegment) int {
+	for i, seg := range e.ess {
+		if seg.Equal(segment) {
+			return i
+		}
+	}
+
+	return -1
+}
+
+func (e sortInterfaceWrapper) strippedOffsetsFor(segmentIndex int) (int64, []int64) {
+	var offsets = make([]int64, 0, e.numerators[segmentIndex]+1)
+	var chosenCounts = make([]int64, len(e.ess))
+	// Here instead of calculating steps which need to be big.Rat, we use the fact that
+	// the steps are always the length of the segment inverted which also is lcd/numerator
+	// So instead of creating and adding up big.Rat we just multiply the step by the amount
+	// of times given segment has been chosen which is count * lcd / numerator and use that
+	// this both saves on a lot of big.Rat allocations and also on a lot of unneeded calculations
+	// with them.
+
+	for i := int64(0); i < e.lcd; i++ {
+		for index, chosenCount := range chosenCounts {
+			num := chosenCount * e.lcd
+			denom := e.numerators[index]
+			if i > num/denom || (i == num/denom && num%denom == 0) {
+				chosenCounts[index]++
+				if index == segmentIndex {
+					prev := int64(0)
+					if len(offsets) > 0 {
+						prev = offsets[len(offsets)-1]
+					}
+					offsets = append(offsets, i-prev)
+					if int64(len(offsets)) == e.numerators[index] {
+						offsets = append(offsets, offsets[0]+e.lcd-i)
+						return offsets[0], offsets[1:]
+					}
+				}
+				break
```

Comment:
I find this difficult to follow, but probably because I don't have an intuitive understanding of the algorithm yet. Nice work!

Reply:
Even though we discussed this, I also found it a bit difficult to follow, but once I grokked it, I quite like how cleverly you've implemented it, so big :+1: That middle part in `if index == segmentIndex { <this code> }` might be a bit easier to understand with a simple `saveIndex(int) bool` lambda that returns if the loop should break. Then you also wouldn't need to panic at the end of the function.